python/lsst/daf/butler/registry/interfaces/_versioning.py

```diff
@@ -108,3 +155,114 @@ def extensionName(cls) -> str:
             Full extension name.
         """
         return f"{cls.__module__}.{cls.__name__}"
+
+    @classmethod
+    @abstractmethod
+    def currentVersion(cls) -> VersionTuple | list[VersionTuple] | None:
+        """Return schema version(s) supported by this extension class.
+
+        Returns
+        -------
+        version : `VersionTuple`, `list` [`VersionTuple`], or `None`
+            Schema version for this extension. `None` is returned if an
+            extension does not require its version to be saved or checked.
+            Extensions that support multiple schema versions return the list
+            of versions.
+        """
+        raise NotImplementedError()
+
+    def newSchemaVersion(self) -> VersionTuple | None:
+        """Return schema version for newly created registry.
+
+        Returns
+        -------
+        version : `VersionTuple` or `None`
+            Schema version created by this extension. `None` is returned if an
+            extension does not require its version to be saved or checked.
+
+        Notes
+        -----
+        Default implementation only forks for extensions that support single
+        schema version and it returns version obtained from `currentVersion`.
+        If `currentVersion` returns multiple version then default
+        implementation will raise an exception and the method has to be
+        reimplemented by a subclass.
+        """
+        my_version = self.currentVersion()
+        if my_version is None or isinstance(my_version, VersionTuple):
+            return my_version
+        raise NotImplementedError(
+            f"Extension {self.extensionName()} supports multiple schema versions, "
+            "its newSchemaVersion() method needs to be re-implemented."
+        )
+
+    @classmethod
+    def checkCompatibility(cls, registry_schema_version: VersionTuple, update: bool) -> None:
+        """Check that schema version defined in registry is compatible with
+        current implementation.
+
+        Parameters
+        ----------
+        registry_schema_version : `VersionTuple`
+            Schema version that exists in registry or defined in a
+            configuration for a registry to be created.
+        update : `bool`
+            If True then read-write access is expected.
+
+        Raises
+        ------
+        IncompatibleVersionError
+            Raised if schema version is not supported by implementation.
+
+        Notes
+        -----
+        Default implementation uses `VersionTuple.checkCompatibility` on
+        the version(s) returned from `currentVersion` method. Subclasses that
+        support different compatibility model will overwrite this method.
+        """
+        # Extensions that do not define current versions are compatible with
+        # anything.
+        if (my_version := cls.currentVersion()) is not None:
+            if isinstance(my_version, VersionTuple):
```

Comment:
are we sure that `currentVersion` can't always return a list so that we could remove this check?

Reply:
Well, it certainly can, I just did not want to touch all existing implementations of `currentVersion`. Probably worth doing that though, to avoid checking for type (and then rename it to `currentVersions`).

Reply:
I changed its name to `currentVersions` and it now returns `list[VersionTuple]`.